libbpf-tools/drsnoop.bpf.c

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include "drsnoop.h"
+
+#define BPF_F_INDEX_MASK		0xffffffffULL
+#define BPF_F_CURRENT_CPU		BPF_F_INDEX_MASK
+
+const volatile pid_t targ_pid = 0;
+const volatile pid_t targ_tgid = 0;
+const __u64 * const volatile vm_zone_stat_kaddr = 0;
+
+struct piddata {
+	u64 ts;
+	u64 nr_free_pages;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u32);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+SEC("tp_btf/mm_vmscan_direct_reclaim_begin")
+int handle__mm_vmscan_direct_reclaim_begin(u64 *ctx)
+{
+	u64 vm_zone_stat[NR_VM_ZONE_STAT_ITEMS];
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+	if (targ_pid && targ_pid != pid)
+		return 0;
+
+	piddata.ts = bpf_ktime_get_ns();
+	if (vm_zone_stat_kaddr) {
+		bpf_probe_read(&piddata.nr_free_pages,
+			       sizeof(*vm_zone_stat_kaddr),
+			       (const void*)&vm_zone_stat_kaddr[NR_FREE_PAGES]);
```

Comment:
nit: you shouldn't need a cast here, does compiler complain without it?

Reply:
No compiler complain without it. It's my fault.